File: guess_number_high_low.cpp
#include "leetcode.h"

class Solution {
public:
  int guessNumber(int n) {
    int left = 1, right = n, mid = n / 2;
    while (left <= right) {
      int ans = guess(mid); // guess is api call
      if (!ans)
        return mid;
      left = ans < 0 ? left : mid + 1;
      right = ans < 0 ? mid - 1 : right;
      mid = left + (right - left) / 2;
    }
    return -1;
  }
};

int main() {
  Solution obj;
  cout << obj.guessNumber(10);
}

File: guess_number_high_low.rs
struct Solution;

impl Solution {
    unsafe fn guessNumber(n: i32) -> i32 {
        let (mut s, mut n) = (1, n as i64);
        while s <= n {
            let x = (n + s) / 2;
            let g = guess(x as i32);
            match g {
                0 => return x as i32,
                1 => s = x + 1,
                -1 => n = x - 1,
                _ => unreachable!()
            }
        }
        unreachable!()
    }
}

fn main() {
    print!("{}", Solution::guessNumber(10));
}
